File: 6. Zigzag Conversion/6. Zigzag Conversion.java
import java.util.ArrayList;
import java.util.List;

class Solution {

  public String convert(String s, int numRows) {
    if (numRows == 1) {
      return s;
    }

    List<StringBuilder> rows = new ArrayList<>();
    for (int i = 0; i < numRows; i++) {
      rows.add(new StringBuilder());
    }

    int pos = 0;
    int dPos = 1;

    for (char c : s.toCharArray()) {
      rows.get(pos).append(c);

      if (pos == numRows - 1) {
        dPos = -1;
      } else if (pos == 0) {
        dPos = 1;
      }

      pos += dPos;
    }

    StringBuilder zigZag = new StringBuilder();
    for (StringBuilder row : rows) {
      zigZag.append(row);
    }

    return zigZag.toString();
  }
}

File: 6. Zigzag Conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s

        n = len(s)
        res = []
        increment = 2 * (numRows - 1)

        for r in range(numRows):
            for i in range(r, n, increment):
                res.append(s[i])
                if 0 < r < numRows - 1 and i + increment - 2 * r < n:
                    res.append(s[i + increment - 2 * r])

        return "".join(res)
